### PR TITLE
Feature/UPPSF-981 Add PAC Annotations (again)

### PR DESCRIPTION
**Pull request by ivan-p-nikolov**

### CODE REVIEW

File: dispatch/dispatcher.go [hunk 1]
```diff
@@ -70,7 +70,7 @@ func (d *dispatcher) forwardToSubscribers(notification Notification) {
 	var sent, failed, skipped int
```
Comment (MiroslavGatsanoga) [inline dispatch/dispatcher.go]: The acquiring of the lock above is for reading purposes, but at the end of the method we are modifying the state of the dispatcher's history. Semantically it will be better to use the full lock in this method.

File: resources/push.go [hunk 2]
```diff
@@ -159,16 +165,16 @@ func getClientAddr(r *http.Request) string {
...
-	if contentType == "" {
+	subType := r.URL.Query().Get("type")
+	if subType == "" {
 		return defaultSubscriptionType, nil
 	}
-	for _, t := range supportedContentTypes {
-		if strings.ToLower(contentType) == strings.ToLower(t) {
-			return contentType, nil
+	for _, t := range supportedSubscriptionTypes {
+		if strings.ToLower(subType) == strings.ToLower(t) {
```
Comment (epavlova) [inline resources/push.go]: One interesting suggestion from my linters here is the following function - https://golang.org/pkg/strings/#EqualFold. I thought it's worth sharing.

File: dispatch/notifications.go [hunk 3]
```diff
@@ -1,17 +1,33 @@
...
-	PublishReference string   `json:"publishReference,omitempty"`
-	LastModified     string   `json:"lastModified,omitempty"`
-	NotificationDate string   `json:"notificationDate,omitempty"`
-	Title            string   `json:"title,omitempty"`
-	Standout         Standout `json:"standout"`
-	ContentType      string   `json:"-"`
+	APIURL           string    `json:"apiUrl"`
+	ID               string    `json:"id"`
+	Type             string    `json:"type"`
+	SubscriberId     string    `json:"subscriberId,omitempty"`
```
Comment (epavlova) [inline dispatch/notifications.go]: I'm not sure why the linters are passing for this one: `SubscriberId` should be `SubscriberID`.

File: consumer/consume_test.go [hunk 4]
```diff
@@ -27,7 +27,7 @@ func TestSyntheticMessage(t *testing.T) {
 	}
 
 	dispatcher := new(mocks.MockDispatcher)
-	handler := NewMessageQueueHandler(defaultContentUriWhitelist, NewSet(), mapper, dispatcher)
+	handler := NewContentQueueHandler(defaultContentUriWhitelist, NewSet(), mapper, dispatcher)
```
Comment (epavlova) [inline consumer/consume_test.go]: It looks to me that this file `consume_test.go` has tests related to `content.go` and is more appropriate to be `content_test.go`.

File: consumer/mapper_test.go [hunk 5]
```diff
@@ -3,7 +3,7 @@ package consumer
 import (
 	"testing"
 
-	"github.com/satori/go.uuid"
+	uuid "github.com/satori/go.uuid"
```
Comment (epavlova) [inline consumer/mapper_test.go]: There are no tests for `MapMetadataNotification`.

File: consumer/metadata.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+		originWhitelist: originWhitelist,
+		mapper:          mapper,
+		dispatcher:      dispatcher,
+	}
+}
+
+func (h *metadataQueueHandler) HandleMessage(queueMsg kafka.FTMessage) error {
+	msg := NotificationQueueMessage{queueMsg}
+	tid := msg.TransactionID()
+	logger := log.WithTransactionID(tid)
```
Comment (epavlova) [inline consumer/metadata.go]: One much broader topic that is not part of this PR: do we want to include the _sending of push notification for annotation changes_ as a mandatory step for publishing annotations. If the answer is true we should change the monitoring for publishing annotations and include it there. Then the logger here should look like the logger in `content.go`:
```
monitoringLogger := log.WithMonitoringEvent("NotificationsPush", tid, contentType)
```

File: consumer/metadata_test.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+	logger.InitDefaultLogger("test-notifications-push")
+
+	tests := map[string]struct {
+		mapper      NotificationMapper
+		sendFunc    func(notification []dispatch.Notification)
+		whitelist   []string
+		msg         kafka.FTMessage
+		expectError bool
+	}{
+		"Test fail on ivalid message body": {
```
Comment (epavlova) [inline consumer/metadata_test.go]: A typo - `invalid`.

File: consumer/content.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+type contentQueueHandler struct {
+	contentUriWhitelist  *regexp.Regexp
+	contentTypeWhitelist *Set
+	mapper               NotificationMapper
+	dispatcher           dispatch.Dispatcher
+}
+
+// NewContentQueueHandler returns a new message handler
+func NewContentQueueHandler(contentUriWhitelist *regexp.Regexp, contentTypeWhitelist *Set, mapper NotificationMapper, dispatcher dispatch.Dispatcher) *contentQueueHandler {
```
Comment (epavlova) [inline consumer/content.go]: Here again we have something strange with the linters - `contentUriWhitelist` should be `contentURIWhitelist`.

File: consumer/content.go [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+	s.m[value] = exists
+}
+
+func (s *Set) Contains(value string) bool {
+	_, c := s.m[value]
+	return c
+}
+
+type contentQueueHandler struct {
+	contentUriWhitelist  *regexp.Regexp
```
Comment (epavlova) [inline consumer/content.go]: Here again we have something strange with the linters - `contentUriWhitelist` should be `contentURIWhitelist`.

File: resources/push.go (hunk 2 above)
Comment (ivan-p-nikolov) [reply resources/push.go]: Will address in code improvements PR

File: dispatch/notifications.go (hunk 3 above)
Comment (ivan-p-nikolov) [reply dispatch/notifications.go]: Will address in code improvements PR.

File: consumer/content.go (hunk 9 above)
Comment (ivan-p-nikolov) [reply consumer/content.go]: Will address in code improvements PR.

File: consumer/content.go (hunk 8 above)
Comment (ivan-p-nikolov) [reply consumer/content.go]: Will address in code improvements PR.

File: consumer/mapper_test.go (hunk 5 above)
Comment (ivan-p-nikolov) [reply consumer/mapper_test.go]: It has a non explicit one in metadata_test "Test Handle Message". I can add a new one in mapper_test as well

File: app_integration_test.go [hunk 10]
```diff
@@ -231,5 +231,5 @@ func createMsgQueue(t *testing.T, uriWhitelist string, typeWhitelist []string, r
 		Resource:   resource,
 		APIBaseURL: apiURL,
 	}
-	return consumer.NewMessageQueueHandler(reg, set, mapper, d)
+	return consumer.NewContentQueueHandler(reg, set, mapper, d)
```
[the remark this answers is not in the document]
Comment (ivan-p-nikolov) [reply app_integration_test.go]: Added new test case for annotations.
Because new logger integration is partial, meaning half of the packages are using the old logger and half the new, we need to initialize the old global logger as well.

[the remark this answers is not in the document]
Comment (epavlova) [reply app_integration_test.go]: 👍

File: dispatch/notifications.go (hunk 3 above)
Comment (epavlova) [reply dispatch/notifications.go]: 👍

File: consumer/content.go (hunk 8 above)
Comment (epavlova) [reply consumer/content.go]: 👍

File: consumer/content.go (hunk 9 above)
Comment (epavlova) [reply consumer/content.go]: 👍 I'm still wondering why the linters are not working :(

File: resources/push.go (hunk 2 above)
Comment (epavlova) [reply resources/push.go]: 👍
